{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ILLUMINA PAIRED-END READ SEQUENCING SIMULATOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "import bisect\n",
    "import random\n",
    "import os\n",
    "import numpy\n",
    "import numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleotids = ['A', 'C', 'G', 'T']\n",
    "complNucleotids = {\n",
    "  \"A\": \"T\",\n",
    "  \"T\": \"A\",\n",
    "  \"G\": \"C\", \n",
    "  \"C\": \"G\"\n",
    "}\n",
    "\n",
    "insertionPositions = []\n",
    "\n",
    "# Mask for SNV formated so the last 2 bits can be used for nucleotid index in nucleotids list\n",
    "SNV = 0b100\n",
    "DEL = 'D'\n",
    "\n",
    "# Helper function for generating most optimal standard deviation for given mean.\n",
    "# This is necessary because we are taking values from a speceific range of normal distribution.\n",
    "def getStandardDeviation(averageQuality):\n",
    "    if(averageQuality < 40 or averageQuality > 120):\n",
    "        return 1\n",
    "    elif(averageQuality < 50 or averageQuality > 105):\n",
    "        return 2\n",
    "    elif(averageQuality < 70 or averageQuality > 90):\n",
    "        return 5\n",
    "    else:\n",
    "        return 10\n",
    "\n",
    "# Generates qualities for nucleotids for one read using normal distribution and writes them in the given file \n",
    "def generateQuality(averageQuality, readSize):\n",
    "    sigma = getStandardDeviation(averageQuality)\n",
    "    a, b = (33 - averageQuality) / sigma, (126 - averageQuality) / sigma\n",
    "    quality = stats.truncnorm.rvs(a, b, scale=sigma, loc=averageQuality, size=readSize)\n",
    "    return \"\".join(map(lambda q: chr(int(round(q))), quality))\n",
    "\n",
    "# Calculates necessary number of fragments based on the given coverage\n",
    "def numOfFragments(coverage, genomeSize, readSize):\n",
    "    return int(round((coverage * genomeSize) / (2 * readSize)))\n",
    "\n",
    "# Finds the 1-based leftmost position where read matches refGenome.\n",
    "def getLeftmostPosition(refGenome, startingPosition, endingPosition):\n",
    "    pos = startingPosition\n",
    "    while((pos <= endingPosition) and isinstance(refGenome[pos], int) and  not(refGenome[pos] & SNV)):\n",
    "        pos += 1\n",
    "    return pos + 1 - bisect.bisect_right(insertionPositions, pos);\n",
    "\n",
    "# Generate one read. Direction defines should it be read from left to right or vice-versa.\n",
    "def generateSingleRead(readSize, refGenome, refPos, direction, complFunction):\n",
    "    j = 0;\n",
    "    read = []\n",
    "    while (len(read) < readSize):\n",
    "        if(isinstance(refGenome[refPos], str)):\n",
    "            if(refGenome[refPos] == DEL):\n",
    "                refPos += direction\n",
    "            else:\n",
    "                read.append(complFunction(refGenome[refPos]))\n",
    "                refPos += direction\n",
    "        elif(refGenome[refPos] & SNV):\n",
    "            read.append(complFunction(nucleotids[refGenome[refPos] & 3]))\n",
    "            refPos += direction\n",
    "        else:\n",
    "            read.append(complFunction(nucleotids[refGenome[refPos] & 3]))\n",
    "            refPos += direction\n",
    "    return (\"\".join(read), refPos - direction)\n",
    "\n",
    "# Genetares 2 FASTQ files which contain paired-end reads.\n",
    "def generateReads(refGenomeDict, quality, coverage, readSize, insertSize, fileName):\n",
    "    samFile = open(fileName + \".sam\", \"w\")\n",
    "    for refGenomeName, refGenome in refGenomeDict.items():\n",
    "        samFile.write(\"@SQ SN:{} LN:{}\\n\".format(refGenomeName, len(refGenome)))\n",
    "        \n",
    "    read1File = open(fileName + \"_read1.fastq\",\"w\")\n",
    "    read2File = open(fileName + \"_read2.fastq\", \"w\")\n",
    "    \n",
    "    for refGenomeName, refGenome in refGenomeDict.items():\n",
    "        genomeSize = len(refGenome)\n",
    "        fragmentNumber = numOfFragments(coverage, genomeSize, readSize)\n",
    "        for i in range(fragmentNumber):\n",
    "            # Insert size follows a normal distribution so we are simulating that\n",
    "            recalInsertSize = int(numpy.random.normal(insertSize, insertSize * 0.05))\n",
    "            # Randomly choose position for the next fragment.\n",
    "            fragmentPosition = genomeSize\n",
    "            while (fragmentPosition + recalInsertSize >= genomeSize):\n",
    "                fragmentPosition = int(random.random()*(genomeSize - recalInsertSize + 1))\n",
    "            \n",
    "            readId = (\"{}_{}_{}_{}\".format(refGenomeName, (fragmentPosition +1), (fragmentPosition + recalInsertSize + 1), i))\n",
    "            \n",
    "            # Generate first paried-end read.\n",
    "            refPos = fragmentPosition\n",
    "            readAndLastIndex = generateSingleRead(readSize, refGenome, refPos, 1, lambda x: x)\n",
    "            qualityOfRead = generateQuality(quality, readSize)\n",
    "            read1File.write(\"@{}/1\\n{}\\n+\\n{}\\n\".format(readId, readAndLastIndex[0], qualityOfRead))\n",
    "\n",
    "            leftmostPosition = getLeftmostPosition(refGenome, refPos, readAndLastIndex[1])\n",
    "            samFile.write(\"{} {} {} {}\\n\".format(readId, leftmostPosition, readAndLastIndex[0], qualityOfRead))\n",
    "            \n",
    "            # Generate second paried-end read.\n",
    "            refPos = fragmentPosition + recalInsertSize\n",
    "            readAndFirstIndex = generateSingleRead(readSize, refGenome, refPos, -1, lambda x: complNucleotids[x])\n",
    "            qualityOfRead = generateQuality(quality, readSize)\n",
    "            read2File.write(\"@{}/2\\n{}\\n+\\n{}\\n\".format(readId, readAndFirstIndex[0], qualityOfRead)) \n",
    "            originalRead = []\n",
    "            i = len(readAndFirstIndex[0]) - 1\n",
    "            while(i > -1):\n",
    "                originalRead.append(complNucleotids[readAndFirstIndex[0][i]])\n",
    "                i-=1\n",
    "            leftmostPosition = getLeftmostPosition(refGenome, readAndFirstIndex[1], refPos)        \n",
    "            samFile.write(\"{} {} {} {}\\n\".format(readId, leftmostPosition, \"\".join(originalRead), qualityOfRead))\n",
    "\n",
    "    read1File.close()\n",
    "    read2File.close()\n",
    "    samFile.close()\n",
    "\n",
    "# Generates mutations by inserting them in referenceGenome.\n",
    "def insertMutations(refGenomeDict, errorSNV, errorInDel):\n",
    "    for refGenome in refGenomeDict.values():\n",
    "        # Insert SNV.\n",
    "        numOfSNV = round(len(refGenome)*errorSNV)\n",
    "        while(numOfSNV > 0):\n",
    "            errorPos = -1\n",
    "            while(errorPos < 0 or isinstance(refGenome[errorPos], int)):\n",
    "                errorPos = int(random.random()*len(refGenome))\n",
    "            refGenome[errorPos] = ((nucleotids.index(refGenome[errorPos]) + random.randint(1,3))%4) | SNV\n",
    "            numOfSNV -= 1\n",
    "        \n",
    "        # Insert INDEL.\n",
    "        numOfInDel = round(len(refGenome)*errorInDel)\n",
    "        while(numOfInDel > 0):\n",
    "            errorPos = -1\n",
    "            while(errorPos < 0 or isinstance(refGenome[errorPos], int) or refGenome[errorPos] == 'D'):\n",
    "                errorPos = int(random.random()*len(refGenome))\n",
    "            if (random.random() < 0.5): #Simulate insertion\n",
    "                 # Randomly choose position from nucleotid list and add it in refGenome.\n",
    "                refGenome.insert(errorPos, random.randrange(4))\n",
    "                insertionPositions.append(errorPos)\n",
    "            else: # Simulate deletion.\n",
    "                # Replace nucleotid with mask for deletion.\n",
    "                refGenome[errorPos] = DEL\n",
    "            numOfInDel -= 1\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loads genome from a file.\n",
    "def readGenome(fileName):\n",
    "    refGenome = []\n",
    "    refFile = open(fileName)\n",
    "    refGenomeName = \"\"\n",
    "    refGenomeDict = {}\n",
    "    for line in refFile.readlines():\n",
    "        # Since one FASTA file can contain multiple sequences, we are loading them all and saving them in dict.\n",
    "        if line[0] == '>':\n",
    "            if refGenome:\n",
    "                refGenomeDict[refGenomeName] = refGenome[:]\n",
    "            refGenome = []\n",
    "            refGenomeName = line[1:].rstrip('\\n')\n",
    "        else:\n",
    "            refGenome.extend(list(line.rstrip('\\n')))\n",
    "    if refGenome:\n",
    "        refGenomeDict[refGenomeName] = refGenome[:]    \n",
    "    refFile.close()\n",
    "    return refGenomeDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkPositiveIntValidity(value, name):\n",
    "    if(not isinstance(value, int) or value < 0):\n",
    "        print(\"{} must be whole number bigger then 0. Please choose the correct value and try again.\".format(name))\n",
    "        return 1\n",
    "    return 0\n",
    "\n",
    "def checkProbabilityValidity(value, name):\n",
    "    if(not isinstance(value, numbers.Real) or value < 0 or value > 1):\n",
    "        print(\"{} must represent probability between 0 and 1. Please choose the correct value and try again.\".format(name))\n",
    "        return 1\n",
    "    return 0\n",
    "\n",
    "def validateParameters(quality, coverage, readSize, insertSize, errorSNV, errorInDel):\n",
    "    score = 0;\n",
    "    if(not isinstance(quality, int) or quality < 26 or quality > 133):\n",
    "        print(\"Quality must be whole number between 33 and 126. Please choose the correct value and try again\")\n",
    "        score = 1\n",
    "    if(checkPositiveIntValidity(coverage, \"Coverage\") \n",
    "       | checkPositiveIntValidity(readSize, \"ReadSize\") \n",
    "       | checkPositiveIntValidity(insertSize, \"InsertSize\")):\n",
    "        score = 1\n",
    "    if(readSize > insertSize):\n",
    "        print(\"ReadSize cannot be longer than InsertSize. Please choose the correct value and try again.\")\n",
    "        score = 1\n",
    "    if(checkProbabilityValidity(errorSNV, \"ErrorSNV\") | checkProbabilityValidity(errorInDel, \"ErrorINDEL\")):\n",
    "        score = 1\n",
    "    if(errorSNV + errorInDel > 1):\n",
    "        print(\"Sum of error rates for SNV  and INDEL cannot be larger than 1. Please choose the correct values and try again.\")\n",
    "        score = 1\n",
    "    return score == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main simulation function.\n",
    "def simulatePairedEndSequencing(refGenomeFile, quality, coverage, readSize, insertSize, errorSNV = 0, errorInDel = 0):\n",
    "    isValid = validateParameters(quality, coverage, readSize, insertSize, errorSNV, errorInDel)\n",
    "    if(isValid):\n",
    "        try:\n",
    "            fileNameBase = os.path.basename(refGenomeFile)\n",
    "            fileName = os.path.splitext(fileNameBase)[0]\n",
    "            refGenomeDict = readGenome(refGenomeFile)\n",
    "            # Check if we need to simulate errors.\n",
    "            if(errorSNV + errorInDel > 0):\n",
    "                insertMutations(refGenomeDict, errorSNV, errorInDel)\n",
    "            generateReads(refGenomeDict, quality, coverage, readSize, insertSize, fileName)\n",
    "        except FileNotFoundError:\n",
    "            print(\"File not found. Please check the path and the name and try again.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Simulator is used by calling \"simulatePairedEndReads\". The parameters are:**\n",
    "    - refGenomeFile - path to the FASTA file with reference genome\n",
    "    - quality - average quality of nucleotids. It should be value between 33 and 126\n",
    "    - coverage - Wanted coverage\n",
    "    - readSize - Size of a single read\n",
    "    - insertSize - Size of a fragment\n",
    "    - [OPTIONAL] errorSNV(=0) - Error probability for SNV. Sum of this and errorInDel must not be larger than 1.\n",
    "    - [OPTIONAL] errorInDel(=0) - Error probability for INS/DEL. Sum of this and errorSNV must not be larger than 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
