{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import stats\n",
    "from statistics import mean \n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSigma(averageQuality):\n",
    "    if(averageQuality < 40 or averageQuality > 110):\n",
    "        return 2\n",
    "    elif(averageQuality < 50 or averageQuality > 100):\n",
    "        return 5\n",
    "    elif(averageQuality < 70 or averageQuality > 90):\n",
    "        return 10\n",
    "    else:\n",
    "        return 15\n",
    "\n",
    "def generateQuality(averageQuality, readSize, readFile):\n",
    "    sigma = getSigma(averageQuality)\n",
    "    a, b = (33 - mu) / sigma, (126 - mu) / sigma\n",
    "    for i in range(readSize):\n",
    "        quality = (int(round(stats.truncnorm.rvs(a, b, scale=sigma, loc=mu))))\n",
    "        readFile.write(quality)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleotids = ['A', 'C', 'G', 'T']\n",
    "complNucleotids = {\n",
    "  \"A\": \"T\",\n",
    "  \"T\": \"A\",\n",
    "  \"G\": \"C\", \n",
    "  \"C\": \"G\"\n",
    "}\n",
    "    \n",
    "def normalDistributionNumbersGenerator:\n",
    "\n",
    "def numOfFragments(coverage, genomSize, readLength):\n",
    "    return (coverage * genomSize) / (2 * readLength) #ovde nam je pisalo insertSize ali na netu nadjoh da je readLength, proveri formulu\n",
    "\n",
    "def generateReads(refGenom, quality, coverage, readSize, insertSize):\n",
    "    genomSize = len(refGenom)\n",
    "    fragmentNumber = numOfFragments(coverage, genomSize, readLength)\n",
    "    read1File = open(\"Read1.fastq\",\"w\")\n",
    "    read2File = open(\"Read2.fastq\", \"w\")\n",
    "    for i in fragmentNumber:\n",
    "        fragmentPosition = -1\n",
    "        while (fragmentPosition < 0 and (fragmentPosition + insertsize) > genomSize):\n",
    "            fragmentPosition = normalDistributionNumbersGenerator\n",
    "        j = 0;\n",
    "        while (j < readSize):\n",
    "            refPos = fragmentPosition\n",
    "            if (refGenom[refPos] == 'D'):\n",
    "                refPos += 1\n",
    "            elif (refGenom[refPos] & INS): #radimo inserciju\n",
    "                read[j] = nucleotids[refGenom[refPos] & 3]\n",
    "                j +=1\n",
    "            elif (refGenom[refPos] & SNIP): #radimo zamenu\n",
    "                read[j] = nucleotids[refGenom[refPos] & 3]\n",
    "                j +=1\n",
    "                refPos += 1\n",
    "            else:\n",
    "                read[j] = refGenom[refPos]\n",
    "                j +=1\n",
    "                refPos += 1\n",
    "        \n",
    "        read2File.write(readId) #dodati generisanje readId-a           \n",
    "        read1File.write(read) \n",
    "        read1File.write(\"\\n+\\n\")\n",
    "        generateQuality(quality, readSize, read1File)\n",
    "        \n",
    "        j = 0\n",
    "        while (j < readSize):\n",
    "            refPos = fragmentPosition + insertSize\n",
    "            if(refGenom[refPos] == D):\n",
    "                refPos -= 1\n",
    "            elif(refGenom[refPos] & INS): #radimo inserciju\n",
    "                read[j] = complNucleotids[nucleotids[refGenom[refPos] & 3]]\n",
    "                j += 1\n",
    "            elif(refGenom[refPos] & SNIP): #radimo zamenu\n",
    "                read[j] = complNucleotids[nucleotids[refGenom[refPos] & 3]]\n",
    "                j +=1\n",
    "                refPos -= 1\n",
    "            else:\n",
    "                read[j] = complNucleotids[refGenom[refPos]]\n",
    "                j +=1\n",
    "                refPos -= 1\n",
    "                         \n",
    "        read2File.write(readId)            \n",
    "        read2File.write(read)\n",
    "        read2File.write(\"\\n+\\n\")               \n",
    "        generateQuality(quality, readSize, read2File)\n",
    "    \n",
    "    read1File.close()\n",
    "    read2File.close()\n",
    "\n",
    "def insertMutations(refGenom, errorSnip, errorInDel):\n",
    "    for nucleotids in refGenom:\n",
    "        errorProb = random() # Random float:  0.0 <= x < 1.0\n",
    "        if (errorProb <= (errorSnip + errorInDel)):\n",
    "            if(errorProb <= errorSnip): #radi snip\n",
    "                refGenom[i] = randrange(4) | 4 #randrange -> Integer from 0 to 3 inclusive\n",
    "            else: #radi indel\n",
    "                if (random() < 0.5): #radi inserciju\n",
    "                    refGenom[i] = randrange(4)\n",
    "                else: #radi deleciju\n",
    "                    refGenom[i] = 'D'\n",
    "\n",
    "def createPairEndPairs(refGenom, quality, coverage, readSize, insertSize, errorSnip, errorInDel):\n",
    "    insertMutations(refGenom, errorSnip, errorInDel)\n",
    "    generateReads(refGenom, quality, coverage, readSize, insertSize)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(7& 4 ):\n",
    "    print('jes')\n",
    "f = open('proba.fasta')\n",
    "file_contents = f.read()\n",
    "print(file_contents)\n",
    "lista = list(file_contents)\n",
    "lista[0] = 4\n",
    "print(lista)\n",
    "if(isinstance(lista[1], str)):\n",
    "    if(lista[1] == 'D'): #delecija\n",
    "    else #bez promene\n",
    "elif(lista[1] & 4): #subs\n",
    "else #insercija\n",
    "    print('to')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
