{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "import random\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSigma(averageQuality):\n",
    "    if(averageQuality < 40 or averageQuality > 110):\n",
    "        return 2\n",
    "    elif(averageQuality < 50 or averageQuality > 100):\n",
    "        return 5\n",
    "    elif(averageQuality < 70 or averageQuality > 90):\n",
    "        return 10\n",
    "    else:\n",
    "        return 15\n",
    "\n",
    "def generateQuality(averageQuality, readSize, readFile):\n",
    "    sigma = getSigma(averageQuality)\n",
    "    a, b = (33 - mu) / sigma, (126 - mu) / sigma\n",
    "    for i in range(readSize):\n",
    "        quality = (int(round(stats.truncnorm.rvs(a, b, scale=sigma, loc=mu))))\n",
    "        readFile.write(quality)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleotids = ['A', 'C', 'G', 'T']\n",
    "complNucleotids = {\n",
    "  \"A\": \"T\",\n",
    "  \"T\": \"A\",\n",
    "  \"G\": \"C\", \n",
    "  \"C\": \"G\"\n",
    "}\n",
    "\n",
    "INS = 0b000\n",
    "SNV = 0b100\n",
    "DEL = 'D'\n",
    "    \n",
    "def normalDistributionNumbersGenerator():\n",
    "    return 0\n",
    "\n",
    "def numOfFragments(coverage, genomSize, readLength):\n",
    "    return (coverage * genomSize) / (2 * readLength)\n",
    "\n",
    "def generateReads(refGenome, quality, coverage, readSize, insertSize, fileName):\n",
    "    genomeSize = len(refGenome)\n",
    "    fragmentNumber = numOfFragments(coverage, genomSize, readLength)\n",
    "    read1File = open(\"Read1.fastq\",\"w\")\n",
    "    read2File = open(\"Read2.fastq\", \"w\")\n",
    "   \n",
    "    for i in range(fragmentNumber):\n",
    "        fragmentPosition = -1\n",
    "        while (fragmentPosition < 0 and (fragmentPosition + insertsize) > genomeSize):\n",
    "            fragmentPosition = normalDistributionNumbersGenerator\n",
    "        \n",
    "        j = 0;\n",
    "        while (j < readSize):\n",
    "            refPos = fragmentPosition\n",
    "            if(isinstance(refGenome[refPos], str)):\n",
    "                if (refGenome[refPos] == DEL):\n",
    "                    refPos += 1\n",
    "                else:\n",
    "                    read[j] = refGenome[refPos]\n",
    "                    j +=1\n",
    "                    refPos += 1\n",
    "            elif (refGenome[refPos] & INS): #radimo inserciju\n",
    "                read[j] = nucleotids[refGenome[refPos] & 3]\n",
    "                j +=1\n",
    "            elif (refGenome[refPos] & SNV): #radimo zamenu\n",
    "                read[j] = nucleotids[refGenome[refPos] & 3]\n",
    "                j +=1\n",
    "                refPos += 1\n",
    "            \n",
    "        readId = fileName + \"_\" + fragmentPosition + \"_\" + (fragmentPosition + insertSize) + \"_0:0:0:0_\" + i\n",
    "        read2File.write(readId + \"/1\") #dodati generisanje readId-a           \n",
    "        read1File.write(read) \n",
    "        read1File.write(\"\\n+\\n\")\n",
    "        generateQuality(quality, readSize, read1File)\n",
    "        \n",
    "        j = 0\n",
    "        while (j < readSize):\n",
    "            refPos = fragmentPosition + insertSize\n",
    "            if(isinstance(refGenome[refPos], str)):\n",
    "                if(refGenome[refPos] == DEL):\n",
    "                    refPos -= 1\n",
    "                else:\n",
    "                    read[j] = complNucleotids[refGenome[refPos]]\n",
    "                    j +=1\n",
    "                    refPos -= 1\n",
    "            elif(refGenome[refPos] & INS): #radimo inserciju\n",
    "                read[j] = complNucleotids[nucleotids[refGenome[refPos] & 3]]\n",
    "                j += 1\n",
    "            elif(refGenome[refPos] & SNV): #radimo zamenu\n",
    "                read[j] = complNucleotids[nucleotids[refGenome[refPos] & 3]]\n",
    "                j +=1\n",
    "                refPos -= 1\n",
    "                   \n",
    "        read2File.write(readId + \"/2\")            \n",
    "        read2File.write(read)\n",
    "        read2File.write(\"\\n+\\n\")               \n",
    "        generateQuality(quality, readSize, read2File)\n",
    "    \n",
    "    read1File.close()\n",
    "    read2File.close()\n",
    "\n",
    "def insertMutations(refGenome, errorSnip, errorInDel):\n",
    "    for nucleotids in refGenome:\n",
    "        errorProb = random() # Random float:  0.0 <= x < 1.0\n",
    "        if (errorProb <= (errorSnip + errorInDel)):\n",
    "            if(errorProb <= errorSnip): #radi SNV\n",
    "                refGenome[i] = randrange(4) | SNV #randrange -> Integer from 0 to 3 inclusive\n",
    "            else: #radi indel\n",
    "                if (random() < 0.5): #radi inserciju\n",
    "                    refGenome[i] = randrange(4) | INS\n",
    "                else: #radi deleciju\n",
    "                    refGenome[i] = DEL\n",
    "\n",
    "def readGenome(fileName):\n",
    "    refGenome = []\n",
    "    refFile = open(fileName)\n",
    "    for line in refFile.readlines():\n",
    "        if line[0] != '>':\n",
    "            refGenome += list(line.rstrip('\\n'))\n",
    "    refFile.close()\n",
    "    return refGenome\n",
    " \n",
    "def checkPositiveIntValidity(value, name):\n",
    "    if(not isinstance(value, int) or value < 0):\n",
    "        print(\"{} must be whole number bigger then 0.\".format(name))\n",
    "        return 1\n",
    "    return 0\n",
    "\n",
    "def checkProbabilityValidity(value, name):\n",
    "    if(not isinstance(value, float) or value < 0 or value > 1):\n",
    "        print(\"{} must represent probability between 0 and 1\".format(name))\n",
    "        return 1\n",
    "    return 0\n",
    "    \n",
    "def createPairEndPairs(refGenomeFile, quality, coverage, readSize, insertSize, errorSnip, errorInDel):\n",
    "    score = 0;\n",
    "    if(not isinstance(quality, int) or quality < 26 or quality > 133):\n",
    "        print(\"Quality must be whole number between 33 and 126.\")\n",
    "        score = 1\n",
    "    if(checkPositiveIntValidity(coverage, \"Coverage\") \n",
    "       | checkPositiveIntValidity(readSize, \"ReadSize\") \n",
    "       | checkPositiveIntValidity(insertSize, \"InsertSize\")):\n",
    "        score = 1\n",
    "    if(readSize > insertSize):\n",
    "        print(\"ReadSize cannot be longer than InsertSize\")\n",
    "        score = 1\n",
    "    if(checkProbabilityValidity(errorSnip, \"ErrorSNV\") | checkProbabilityValidity(errorInDel, \"ErrorINDEL\")):\n",
    "        score = 1\n",
    "    if(errorSnip + errorInDel > 1):\n",
    "        print(\"Sum of error rates for SNV  and INDEL cannot be larger than 1\")\n",
    "        score = 1\n",
    "    \n",
    "    if(score == 0):\n",
    "        try:\n",
    "            fileNameBase = os.path.basename(refGenomeFile)\n",
    "            fileName = os.path.splitext(fileNameBase)[0]\n",
    "            refGenome = readGenome(refGenomeFile)\n",
    "            insertMutations(refGenome, errorSnip, errorInDel)\n",
    "            generateReads(refGenome, quality, coverage, readSize, insertSize, fileName)\n",
    "        except FileNotFoundError:\n",
    "            print(\"File not found. Please check the path and the name and try again.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ReadSize cannot be longer than InsertSize\n"
     ]
    }
   ],
   "source": [
    "createPairEndPairs(\"fajl\", quality = 43, coverage = 12, readSize = 32, insertSize = 20, errorSnip = 0.3, errorInDel = 0.6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "float('v')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
