{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run SequencingSimulator.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test for getStandardDeviation function\n",
    "\n",
    "assert getStandardDeviation(35) == 1\n",
    "assert getStandardDeviation(125) == 1\n",
    "assert getStandardDeviation(45) == 2\n",
    "assert getStandardDeviation(115) == 2\n",
    "assert getStandardDeviation(65) == 5\n",
    "assert getStandardDeviation(95) == 5\n",
    "assert getStandardDeviation(75) == 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test for generateQuality function\n",
    "\n",
    "def calculateSum(readLength, quality, readQuality):\n",
    "    sum = 0\n",
    "    for ch in readQuality:\n",
    "        sum += ord(ch)\n",
    "    sum = (int)(sum/readLength)\n",
    "    return sum\n",
    "    \n",
    "def generateQualityTest(quality, readLength):\n",
    "    readQuality = generateQuality(quality, readLength)\n",
    "    assert len(readQuality) == readLength\n",
    "    sum = calculateSum(readLength, quality, readQuality)\n",
    "    assert sum <= quality + 2\n",
    "    assert sum >= quality - 2\n",
    "\n",
    "readLength = 100\n",
    "quality = 70\n",
    "generateQualityTest(quality, readLength)\n",
    "\n",
    "readLength = 150\n",
    "quality = 50\n",
    "generateQualityTest(quality, readLength)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test for generateSingleRead function and getLeftMostPosition\n",
    "\n",
    "refGenome = ['D','T','C','A','G','T','C','G',0,2]\n",
    "readSize = 4\n",
    "refPos = 2\n",
    "direction = 1\n",
    "readAndIndex = generateSingleRead(readSize, refGenome, refPos, direction, lambda x: x)\n",
    "assert len(readAndIndex[0]) == 4\n",
    "assert readAndIndex[0] == \"CAGT\"\n",
    "assert readAndIndex[1] == 5\n",
    "leftmostPosition = getLeftmostPosition(refGenome, refPos, readAndIndex[1])\n",
    "assert leftmostPosition == 3\n",
    "\n",
    "readSize = 4\n",
    "refPos = 7\n",
    "direction = -1\n",
    "readAndIndex = generateSingleRead(readSize, refGenome, refPos, direction, lambda x: complNucleotids[x])\n",
    "assert len(readAndIndex[0]) == 4\n",
    "assert readAndIndex[0] == \"CGAC\"\n",
    "assert readAndIndex[1] == 4\n",
    "leftmostPosition = getLeftmostPosition(refGenome, refPos, readAndIndex[1])\n",
    "assert leftmostPosition == 8\n",
    "\n",
    "readSize = 5\n",
    "refPos = 0\n",
    "direction = 1\n",
    "readAndIndex = generateSingleRead(readSize, refGenome, refPos, direction, lambda x: x)\n",
    "assert len(readAndIndex[0]) == 5\n",
    "assert readAndIndex[0] == \"TCAGT\"\n",
    "assert readAndIndex[1] == 5\n",
    "leftmostPosition = getLeftmostPosition(refGenome, refPos, readAndIndex[1])\n",
    "assert leftmostPosition == 1\n",
    "\n",
    "readSize = 4\n",
    "refPos = 6\n",
    "direction = 1\n",
    "readAndIndex = generateSingleRead(readSize, refGenome, refPos, direction, lambda x: x)\n",
    "assert len(readAndIndex[0]) == 4\n",
    "assert readAndIndex[0] == \"CGAG\"\n",
    "assert readAndIndex[1] == 9\n",
    "leftmostPosition = getLeftmostPosition(refGenome, refPos, readAndIndex[1])\n",
    "assert leftmostPosition == 7\n",
    "\n",
    "refGenome = ['D',7,6,5,4,0]\n",
    "readSize = 5\n",
    "refPos = 0\n",
    "direction = 1\n",
    "readAndIndex = generateSingleRead(readSize, refGenome, refPos, direction, lambda x: x)\n",
    "assert len(readAndIndex[0]) == 5\n",
    "assert readAndIndex[0] == \"TGCAA\"\n",
    "assert readAndIndex[1] == 5\n",
    "leftmostPosition = getLeftmostPosition(refGenome, refPos, readAndIndex[1])\n",
    "assert leftmostPosition == 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A', 'T', 'C', 'G', 'T', 'T', 'A', 'A', 'T', 'G', 'C', 'C', 'G', 'T', 'A']\n",
      "['D', 'T', 'C', 'G', 'T', 'T', 'A', 'A', 4, 5, 'C', 'C', 'G', 'T', 'A']\n",
      "3\n",
      "['A', 'T', 'G', 'G', 'A', 'C', 'G', 'T', 'T', 'A', 'C', 'G', 'T', 'T', 'C', 'G', 'A']\n",
      "['A', 'T', 'G', 'G', 'A', 'C', 'G', 'T', 'T', 'A', 'C', 7, 'T', 'T', 'C', 7, 'A']\n",
      "4\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "#test for readGenome function and insertMutations function\n",
    "\n",
    "fileName = \"GenomTest.fa\"\n",
    "refGenomeDict = readGenome(fileName)\n",
    "assert len(refGenomeDict) == 2\n",
    "assert (\"sequence1\" in refGenomeDict) == True\n",
    "assert len(refGenomeDict[\"sequence1\"]) == 15\n",
    "assert (\"sequence2\" in refGenomeDict) == True\n",
    "assert len(refGenomeDict[\"sequence2\"]) == 17\n",
    "\n",
    "errorSnv = 0.1\n",
    "errorInDel = 0.05\n",
    "originalRefGenome = readGenome(fileName)\n",
    "insertMutations(refGenomeDict, errorSnv, errorInDel)\n",
    "expectedNumOfMut = (round)((len(refGenomeDict[\"sequence1\"]) + len(refGenomeDict[\"sequence2\"])) * (errorSnv + errorInDel))\n",
    "actualNumOfMut = 0\n",
    "j = 0\n",
    "k = 0\n",
    "i = 0\n",
    "origGenome = list(originalRefGenome.values())\n",
    "refGenome = list(refGenomeDict.values())\n",
    "for genome in refGenome:\n",
    "    ogenome = origGenome[j]\n",
    "    print(ogenome)\n",
    "    print(genome)\n",
    "    while i < len(ogenome):\n",
    "        if (genome[k] != ogenome[i]):\n",
    "            if (genome[k] != 0 and genome[k] != 1 and genome[k] != 2 and genome[k] != 3):\n",
    "                i += 1 \n",
    "            k += 1\n",
    "            actualNumOfMut += 1\n",
    "        else :\n",
    "            k += 1\n",
    "            i += 1\n",
    "    j += 1\n",
    "    print(actualNumOfMut)\n",
    "print(actualNumOfMut)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test for validateParameters function\n",
    "\n",
    "assert validateParameters(55, 5, 9, 10, 0.2, 0.1) == True\n",
    "assert validateParameters(25, 5, 9, 10, 0.2, 0.1) == False\n",
    "assert validateParameters(55, -5, 9, 10, 0.2, 0.1) == False\n",
    "assert validateParameters(55, 5, -9, 10, 0.2, 0.1) == False\n",
    "assert validateParameters(55, 5, 9, -10, 0.2, 0.1) == False\n",
    "assert validateParameters(55, 5, 10, 9, 0.2, 0.1) == False\n",
    "assert validateParameters(55, 5, 9, 10, -0.2, 0.1) == False\n",
    "assert validateParameters(55, 5, 9, 10, 0.2, -0.1) == False\n",
    "assert validateParameters(55, 5, 9, 10, 1.2, 0.1) == False\n",
    "assert validateParameters(55, 5, 9, 10, 0.2, 1.1) == False\n",
    "assert validateParameters(55, 5, 9, 10, 0.8, 0.7) == False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test for simulatePairedEndSequencing function\n",
    "\n",
    "readLength = 7\n",
    "simulatePairedEndSequencing(\"GenomTest.fa\", 70, 3, readLength, 12)\n",
    "samFile = open(\"GenomTest.sam\")\n",
    "i = 0\n",
    "for line in samFile.readlines():\n",
    "    if len(line) != 0:\n",
    "        if i == 0:\n",
    "            assert line == \"@SQ SN:sequence1 LN:15\\n\"\n",
    "            i += 1\n",
    "        elif i == 1:\n",
    "            assert line == \"@SQ SN:sequence2 LN:17\\n\"\n",
    "            i += 1\n",
    "        else:\n",
    "            words = line.split()\n",
    "            assert len(words) == 4\n",
    "            assert len(words[2]) == readLength\n",
    "            assert len(words[3]) == readLength\n",
    "            i += 1\n",
    "assert i == (2 + 6 + 8) #2 - there are 2 sequences, 6  - there are 6 reads from first sequence, 8- there are 8 reads from second sequence\n",
    "samFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
